src/stim/stabilizers/pauli_string_iter.inl

```diff
@@ -0,0 +1,296 @@
+    // 2. c2 = (c1 + 1)
+    // 3. c3 = (~c1 & -~c1) - 1
+    // 4. c4 = c3 >> count_trailing_zeros(cur_perm) + 1
+    // 5. next_perm = c2 | c4;
+    // see https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
+    // These steps need to be updated for a list of integers given by cur_perm.
+    // An equivalent algorithm is
+    // t1 = cur | ((1 << num_zeros) - 1)
+    // t2 = (((1 << (next_zero + 1)) - 1))
+    // t3 = ((1 << (next_zero - num_zeros - 1)) - 1)
+    // next_perm = (t1 ^ t2) | t3
+    // where num_zeros = number of trailing zeros in cur_perm
+    // and next_zero = the next zero after num_zeros
+    // The above operations are simpler to implement for simd_bits as they mostly
+    // involve operations with "ones"-masks and already implemented operations like | and ^.
+    simd_bits<W> t1(cur_perm.num_bits_padded()), t2(cur_perm.num_bits_padded());
+    simd_bits<W> t3(cur_perm.num_bits_padded()), t1_inv(cur_perm.num_bits_padded());
+    uint64_t ones_bit = ~0ULL;
+    t1 = cur_perm;
+    size_t num_zeros = count_trailing_zeros(cur_perm);
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        // Either we have an all all ones or an all zeros integer in which case we
+        // skip, otherwise this word must contain the least significant bit
+        if (cur_perm.u64[w] != ones_bit && cur_perm.u64[w]) {
+            // Use the single word algorithm from above
+            t1.u64[w] = cur_perm.u64[w] | ((1ULL << num_zeros) - 1);
+            // Set all previous words to ones
+            for (size_t old_w = 0; old_w < w; old_w++) {
+                t1.u64[old_w] = ones_bit;
+            }
+            break;
+        }
+    }
+    // invert the bits
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        if (t1.u64[w]) {
+            t1_inv.u64[w] = ~t1.u64[w];
+        } else {
+            // all zeros so set to all ones
+            t1_inv.u64[w] = ones_bit;
+        }
+    }
+    size_t next_zero = count_trailing_zeros(t1_inv);
+    for (size_t w = 0; w < cur_perm.num_u64_padded(); w++) {
+        if (next_zero / 64 == w) {
+            if ((next_zero + 1) % 64 == 0) {
+                t2.u64[w] = ones_bit;
+            } else {
+                t2.u64[w] = ((1ULL << (next_zero + 1 - 64 * w)) - 1);
+            }
+            break;
+        } else {
+            t2.u64[w] = ones_bit;
+        }
+    }
+    size_t word_loc = (next_zero - num_zeros - 1) / 64;
+    for (size_t w = 0; w < word_loc; w++) {
+        t3.u64[w] = ones_bit;
+    }
+    // next_zero - num_zeros - 1 - words * 64 is always < 64, so don't need to worry about overflow like above.
+    t3.u64[word_loc] = (1ULL << (next_zero - num_zeros - 1 - word_loc * 64)) - 1;
+    cur_perm = t1;
+    cur_perm ^= t2;
+    cur_perm |= t3;
+}
+
+template <size_t W>
+bool PauliStringIterator<W>::pair_sat_increment() {
+    // This will overflow for large cur_w.
+    size_t num_terms = static_cast<size_t>(pow(3, cur_w));
```

Comment:
Are you sure this actually returns the right answer for all relevant values? Doubles have 53 bits of precision; not enough for a 64 bit integer. It'd be safer to have a method like

```
uint64_t pow3(uint64_t p) {
  assert(p < 41); 
  uint64_t r = 1;
  if (p & 1) r *= 3
  if (p & 2) r *= 9;
  if (p & 4) r *= 81;
  if (p & 8) r *= 6561;
  if (p & 16) r *= 43046721;
  if (p & 32) r *= 1853020188851841;
  return r;
}
```